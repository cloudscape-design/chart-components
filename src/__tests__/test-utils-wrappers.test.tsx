// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { render } from "@testing-library/react";
import fs from "fs";
import Highcharts from "highcharts";
import path from "path";

import CartesianChart from "../../lib/components/cartesian-chart";
import PieChart from "../../lib/components/pie-chart";
import createWrapper from "../../lib/components/test-utils/dom";

describe("Generate test utils ElementWrapper", () => {
  const importPaths = [
    {
      type: "dom",
      relativePath: "../test-utils/dom/index.ts",
    },
    {
      type: "selectors",
      relativePath: "../test-utils/selectors/index.ts",
    },
  ] as const;

  test.each(importPaths)("$type ElementWrapper matches the snapshot", ({ relativePath }) => {
    const testUtilsPath = path.join(__dirname, relativePath);
    const domWrapper = fs.readFileSync(testUtilsPath, "utf8");
    expect(domWrapper).toMatchSnapshot();
  });
});

test("finds charts by type", () => {
  const highchartsOrNoHighcharts = Math.random() > 0.5 ? Highcharts : null;
  render(<CartesianChart highcharts={highchartsOrNoHighcharts} series={[]} data-testid="c1" />);
  render(<CartesianChart highcharts={highchartsOrNoHighcharts} series={[]} data-testid="c2" />);
  render(<PieChart highcharts={highchartsOrNoHighcharts} series={null} data-testid="p1" />);

  expect(createWrapper().findCartesianHighcharts()).not.toBe(null);
  expect(createWrapper().findCartesianHighcharts('[data-testid="c1"]')).not.toBe(null);
  expect(createWrapper().findCartesianHighcharts('[data-testid="c2"]')).not.toBe(null);
  expect(createWrapper().findAllCartesianHighcharts()).toHaveLength(2);

  expect(createWrapper().findPieHighcharts()).not.toBe(null);
  expect(createWrapper().findPieHighcharts('[data-testid="p1"]')).not.toBe(null);
  expect(createWrapper().findAllPieHighcharts()).toHaveLength(1);
});
