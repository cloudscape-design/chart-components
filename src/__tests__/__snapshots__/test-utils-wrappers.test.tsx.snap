// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generate test utils ElementWrapper > 'dom' ElementWrapper matches the snapshot 1`] = `
"// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ElementWrapper } from "@cloudscape-design/test-utils-core/dom";
import { appendSelector } from "@cloudscape-design/test-utils-core/utils";

export { ElementWrapper };

import CartesianChartWrapper from "./cartesian-chart";
import PieChartWrapper from "./pie-chart";

export { CartesianChartWrapper, PieChartWrapper };

declare module "@cloudscape-design/test-utils-core/dist/dom" {
  interface ElementWrapper {
    /**
     * Returns the wrapper of the first cartesian chart that matches the specified CSS selector.
     * If no CSS selector is specified, returns the wrapper of the first cartesian chart.
     * If no matching chart is found, returns \`null\`.
     *
     * @param {string} [selector] CSS Selector
     * @returns {CartesianChartWrapper | null}
     */
    findCartesianHighcharts(selector?: string): CartesianChartWrapper | null;
    /**
     * Returns the wrapper of the first pie chart that matches the specified CSS selector.
     * If no CSS selector is specified, returns the wrapper of the first pie chart.
     * If no matching chart is found, returns \`null\`.
     *
     * @param {string} [selector] CSS Selector
     * @returns {PieChartWrapper | null}
     */
    findPieHighcharts(selector?: string): PieChartWrapper | null;
    /**
     * Returns an array of cartesian chart wrappers that matches the specified CSS selector.
     * If no CSS selector is specified, returns all cartesian charts inside the current wrapper.
     * If no matching chart is found, returns an empty array.
     *
     * @param {string} [selector] CSS Selector
     * @returns {CartesianChartWrapper[]}
     */
    findAllCartesianHighcharts(selector?: string): Array<CartesianChartWrapper>;
    /**
     * Returns an array of pie chart wrappers that matches the specified CSS selector.
     * If no CSS selector is specified, returns all pie charts inside the current wrapper.
     * If no matching chart is found, returns an empty array.
     *
     * @param {string} [selector] CSS Selector
     * @returns {PieChartWrapper[]}
     */
    findAllPieHighcharts(selector?: string): Array<PieChartWrapper>;
  }
}

ElementWrapper.prototype.findCartesianHighcharts = function (selector) {
  return (this as any).findComponent(getComponentSelector(CartesianChartWrapper, selector), CartesianChartWrapper);
};
ElementWrapper.prototype.findPieHighcharts = function (selector) {
  return (this as any).findComponent(getComponentSelector(PieChartWrapper, selector), PieChartWrapper);
};
ElementWrapper.prototype.findAllCartesianHighcharts = function (selector) {
  return (this as any).findAllComponents(CartesianChartWrapper, selector);
};
ElementWrapper.prototype.findAllPieHighcharts = function (selector) {
  return (this as any).findAllComponents(PieChartWrapper, selector);
};

function getComponentSelector(wrapper: { rootSelector: string }, selector?: string) {
  return selector ? appendSelector(selector, \`.\${wrapper.rootSelector}\`) : \`.\${wrapper.rootSelector}\`;
}

export default function wrapper(root: Element = document.body) {
  if (document && document.body && !document.body.contains(root)) {
    console.warn(
      "[AwsUi] [test-utils] provided element is not part of the document body, interactions may work incorrectly",
    );
  }
  return new ElementWrapper(root);
}
"
`;

exports[`Generate test utils ElementWrapper > 'selectors' ElementWrapper matches the snapshot 1`] = `
"// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ElementWrapper } from "@cloudscape-design/test-utils-core/selectors";
import { appendSelector } from "@cloudscape-design/test-utils-core/utils";

export { ElementWrapper };

import CartesianChartWrapper from "./cartesian-chart";
import PieChartWrapper from "./pie-chart";

export { CartesianChartWrapper, PieChartWrapper };

declare module "@cloudscape-design/test-utils-core/dist/selectors" {
  interface ElementWrapper {
    /**
     * Returns a wrapper that matches charts of the given type with the specified CSS selector.
     * If no CSS selector is specified, returns a wrapper that matches charts.
     *
     * @param {"cartesian" | "pie"} [type] Chart type
     * @param {string} [selector] CSS Selector
     * @returns {CartesianChartWrapper | PieChartWrapper}
     */
    findChart(type: "cartesian", selector?: string): CartesianChartWrapper | null;
    findChart(type: "pie", selector?: string): PieChartWrapper | null;

    /**
     * Returns a multi-element wrapper that matches charts of the given type with the specified CSS selector.
     * If no CSS selector is specified, returns a multi-element wrapper that matches charts.
     *
     * @param {"cartesian" | "pie"} [type] Chart type
     * @param {string} [selector] CSS Selector
     * @returns {MultiElementWrapper<CartesianChartWrapper> | MultiElementWrapper<PieChartWrapper>}
     */
    findAllCharts(type: "cartesian", selector?: string): Array<CartesianChartWrapper>;
    findAllCharts(type: "pie", selector?: string): Array<PieChartWrapper>;
  }
}

ElementWrapper.prototype.findChart = function (type, selector) {
  switch (type) {
    case "cartesian":
      return (this as any).findComponent(getComponentSelector(CartesianChartWrapper, selector), CartesianChartWrapper);
    case "pie":
      return (this as any).findComponent(getComponentSelector(PieChartWrapper, selector), PieChartWrapper);
  }
};

ElementWrapper.prototype.findAllCharts = function (type, selector) {
  switch (type) {
    case "cartesian":
      return (this as any).findAllComponents(CartesianChartWrapper, selector);
    case "pie":
      return (this as any).findAllComponents(PieChartWrapper, selector);
  }
};

function getComponentSelector(wrapper: { rootSelector: string }, selector?: string) {
  return selector ? appendSelector(selector, \`.\${wrapper.rootSelector}\`) : \`.\${wrapper.rootSelector}\`;
}

export default function wrapper(root: string = "body") {
  return new ElementWrapper(root);
}
"
`;
